File: sdk/mobile/ios/Sources/ODPS/Batching.swift
import Foundation

public protocol NetworkStateProvider {
    func isNetworkAvailable() -> Bool
}

public protocol TelemetryUploader {
    func upload(batch: [TransmissionEnvelope]) -> UploadResult
}

public struct UploadResult {
    public let success: Bool
    public let errorMessage: String?

    public init(success: Bool, errorMessage: String? = nil) {
        self.success = success
        self.errorMessage = errorMessage
    }
}

public struct TransmissionEnvelope: Equatable {
    public let event: TelemetryEvent
    public let filteredPayload: [String: AnyHashable]
    public let blockedFields: Set<String>
    public let blockedSensitiveFields: Set<String>
    public let sanitizedMetrics: [Metric]

    public init(event: TelemetryEvent,
                filteredPayload: [String: AnyHashable],
                blockedFields: Set<String>,
                blockedSensitiveFields: Set<String>,
                sanitizedMetrics: [Metric]) {
        self.event = event
        self.filteredPayload = filteredPayload
        self.blockedFields = blockedFields
        self.blockedSensitiveFields = blockedSensitiveFields
        self.sanitizedMetrics = sanitizedMetrics
    }
}

public final class BatchStorage {
    private var queue: [TransmissionEnvelope] = []
    private let lock = NSLock()

    public init() {}

    public func enqueue(_ envelope: TransmissionEnvelope) {
        lock.lock(); defer { lock.unlock() }
        queue.append(envelope)
    }

    public func drain() -> [TransmissionEnvelope] {
        lock.lock(); defer { lock.unlock() }
        let drained = queue
        queue.removeAll()
        return drained
    }

    public func requeue(_ envelopes: [TransmissionEnvelope]) {
        guard !envelopes.isEmpty else { return }
        lock.lock(); defer { lock.unlock() }
        queue = envelopes + queue
    }
}

File: sdk/mobile/android/src/main/kotlin/com/summit/odps/android/Batching.kt
package com.summit.odps.android

fun interface NetworkStateProvider {
  fun isNetworkAvailable(): Boolean
}

fun interface TelemetryUploader {
  fun upload(batch: List<TransmissionEnvelope>): UploadResult
}

data class UploadResult(
  val success: Boolean,
  val errorMessage: String? = null
)

data class TransmissionEnvelope(
  val event: TelemetryEvent,
  val filteredPayload: Map<String, Any?>,
  val blockedFields: Set<String>,
  val blockedSensitiveFields: Set<String>,
  val sanitizedMetrics: List<Metric>
)

class BatchStorage {
  private val queue: ArrayDeque<TransmissionEnvelope> = ArrayDeque()

  @Synchronized
  fun enqueue(envelope: TransmissionEnvelope) {
    queue.addLast(envelope)
  }

  @Synchronized
  fun drain(): List<TransmissionEnvelope> {
    if (queue.isEmpty()) return emptyList()
    val drained = queue.toList()
    queue.clear()
    return drained
  }

  @Synchronized
  fun requeue(items: List<TransmissionEnvelope>) {
    if (items.isEmpty()) return
    val existing = ArrayDeque<TransmissionEnvelope>()
    existing.addAll(items)
    existing.addAll(queue)
    queue.clear()
    queue.addAll(existing)
  }

  @Synchronized
  fun size(): Int = queue.size
}
